In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


ModuleNotFoundError: No module named 'matplotlib'

In [2]:
!pip install matplotlib


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 159.4/159.4 kB 357.7 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 426.6 kB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.4/243.4 kB 410.6 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 367.2 kB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.2/66.2 kB 313.3 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 484.2 kB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.2/103.2 kB 263.1 kB/s eta 0:00:00a 0:00:01


# This neural network is for xor gate

In [3]:
class NeuralNetwork:
    def __init__(self, input_size, hidden_size, output_size):
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        
        # Initialize weights and biases
        self.W1 = np.random.randn(self.input_size, self.hidden_size)
        self.b1 = np.zeros((1, self.hidden_size))
        self.W2 = np.random.randn(self.hidden_size, self.output_size)
        self.b2 = np.zeros((1, self.output_size))
        
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))
    
    def sigmoid_derivative(self, x):
        return x * (1 - x)
    
    def forward(self, X):
        # Forward pass
        self.z1 = np.dot(X, self.W1) + self.b1
        self.a1 = self.sigmoid(self.z1)
        self.z2 = np.dot(self.a1, self.W2) + self.b2
        self.a2 = self.sigmoid(self.z2)
        return self.a2
    
    def backward(self, X, y, output, learning_rate):
        # Backward pass
        error = y - output
        delta_output = error * self.sigmoid_derivative(output)
        error_hidden = np.dot(delta_output, self.W2.T)
        delta_hidden = error_hidden * self.sigmoid_derivative(self.a1)
        
        # Update weights and biases
        self.W2 += np.dot(self.a1.T, delta_output) * learning_rate
        self.b2 += np.sum(delta_output, axis=0, keepdims=True) * learning_rate
        self.W1 += np.dot(X.T, delta_hidden) * learning_rate
        self.b1 += np.sum(delta_hidden, axis=0, keepdims=True) * learning_rate
    
    def train(self, X, y, epochs=200, learning_rate=0.01):
        for epoch in range(epochs):
            # Forward propagation
            output = self.forward(X)
            
            # Backpropagation and weight update
            self.backward(X, y, output, learning_rate)
            
            # Print loss every 10 epochs
            if epoch % 10 == 0:
                loss = np.mean(np.abs(y - output))
                print(f'Epoch {epoch}, Loss: {loss}')


In [4]:
X = np.array([[0, 0],
              [0, 1],
              [1, 0],
              [1, 1]])

# Target output
y = np.array([[0],
              [1],
              [1],
              [0]])
input_size = 2
hidden_size = 4
output_size = 1
nn = NeuralNetwork(input_size, hidden_size, output_size)

# Train the neural network
nn.train(X, y)

Epoch 0, Loss: 0.4988588193551401
Epoch 10, Loss: 0.4988397342262523
Epoch 20, Loss: 0.4988222494255704
Epoch 30, Loss: 0.49880630317887353
Epoch 40, Loss: 0.4987918186860209
Epoch 50, Loss: 0.49877870703593896
Epoch 60, Loss: 0.4987668700763336
Epoch 70, Loss: 0.49875620314736985
Epoch 80, Loss: 0.4987465976098168
Epoch 90, Loss: 0.49873794311903563
Epoch 100, Loss: 0.4987301296154275
Epoch 110, Loss: 0.49872304901870534
Epoch 120, Loss: 0.4987165966271644
Epoch 130, Loss: 0.4987106722338823
Epoch 140, Loss: 0.49870518097959404
Epoch 150, Loss: 0.49870003396715773
Epoch 160, Loss: 0.4986951486654426
Epoch 170, Loss: 0.4986904491315356
Epoch 180, Loss: 0.49868586607981874
Epoch 190, Loss: 0.49868133682508653


In [3]:
import numpy as np
from keras.models import Sequential
from keras.layers import Dense
import keras


# Define the same dataset
X = np.array([[0, 0],
              [0, 1],
              [1, 0],
              [1, 1]])

# Target output
y = np.array([[0],
              [1],
              [1],
              [0]])

# Define the model using Keras
model = Sequential([
    Dense(4, input_shape=(2,), activation='sigmoid'),
    Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
history = model.fit(X, y, epochs=200, verbose=0)

# Evaluate the model
loss = model.evaluate(X, y, verbose=0)
print("Keras Model Loss:", loss)

# Print the loss history of Keras model every 10 epochs
for epoch, loss in enumerate(history.history['loss']):
    if epoch % 10 == 0:
        print(f'Epoch {epoch}, Loss: {loss}')


ModuleNotFoundError: No module named 'tensorflow'

#### Inference

Vanilla NN gave a loss of <b>0.4988</b> but the built in keras library performed better by yeilding a loss of <b>0.2511</b>